File: exemplos/regex.php
<?php

$regex = '~(\d\d)(\w)~'; // Repare nos tils envolvendo a RegEx.
echo "Pattern: $regex" . PHP_EOL;

$alvo = '11a22b33c';
echo "Alvo: $alvo" . PHP_EOL;

echo PHP_EOL;
echo "Execução da função preg_match" . PHP_EOL;

$achou = preg_match($regex, $alvo, $match); // $match recebe só o primeiro match.
// preg_match retorna um booleano, confirmando se houve match ou não.
print_r($match); // Imprime o conteúdo de uma variável de forma legível para humanos.
    /**
     * Resultado esperado:
     * Array
     * (
     *     [0] => 11a
     *     [1] => 11
     *     [2] => a
     * )
     * 
     */
echo "Encontrou algum match? $achou" . PHP_EOL;

echo PHP_EOL;
echo "Execução da função preg_match_all" . PHP_EOL;

$achou = preg_match_all($regex, $alvo, $match); // $match recebe todos os matches.
echo "Encontrou algum match? $achou" . PHP_EOL; // Retorna 3, o número de matches.

print_r($match); // Imprime o conteúdo de uma variável de forma legível para humanos.
    /**
     * Resultado esperado: 
     * Array
     * (
     *     [0] => Array         // Match completo.
     *         (
     *             [0] => 11a
     *             [1] => 22b
     *             [2] => 33c
     *         )
     * 
     *     [1] => Array         // Primeiro grupo.
     *         (
     *             [0] => 11
     *             [1] => 22
     *             [2] => 33
     *         )
     * 
     *     [2] => Array         // Segundo grupo.
     *         (
     *             [0] => a
     *             [1] => b
     *             [2] => c
     *         )
     * 
     * )
     */


echo PHP_EOL;
echo "Capturando os índices no alvo com a função preg_match_all:" . PHP_EOL;
echo "\tForneça o quarto parâmetro PREG_OFFSET_CAPTURE" . PHP_EOL;

$achou = preg_match_all($regex, $alvo, $match, PREG_OFFSET_CAPTURE);
echo "Encontrou algum match? $achou" . PHP_EOL; // Retorna 3, o número de matches.

print_r($match); // Imprime o conteúdo de uma variável de forma legível para humanos.
echo "
    O resultado é um array de 3 dimensões, sendo:
        a primeira dimensão os grupos de matches (numerados de 0 a n), 
        a segunda dimensão as combinações de cada grupo de 0 a n, e
        a terceira dimensão contém a combinação do grupo + a posição no alvo.
    Outra leitura:
        1) dimensão dos grupos (inclusive o grupo zero, da combinação total);
        2) dimensão das combinações dentro do grupo;
        3) dimensão do texto combinado e da posição.
";

File: exemplos/regex.py
regex = r'(\d\d)(\w)' 
    # O r antes das aspas converte a string em uma raw
    # string, que não permite escape de caracteres.

alvo = '11a22b33c'

import re
# Importa a biblioteca de expressões regulares.

resultado = re.search(regex, alvo) # Retorna apenas o primeiro match.
print(resultado)

print(f"Uma forma de mostrar grupo de match inteiro: {resultado.group()}")
print(f"Outra forma de mostrar grupo de match inteiro: {resultado.group(0)}")
print(f"\t1o grupo de match: {resultado.group(1)}")
print(f"\t2o grupo de match: {resultado.group(2)}")

print("\n")

print(f"Posição inicial do match no alvo: {resultado.start()}")
print(f"Posição final do match no alvo: {resultado.end()}")

print("\n")

print("Buscando todos os matches")
resultados = re.finditer(regex, alvo) # Retorna um iterador.
# Uma vez que esse iterador é executado, ele não pode ser reexecutado.
for resultado in resultados:
    print("\t%s | %s | %s [%s,%s]" % (resultado.group(), resultado.group(1), resultado.group(2), resultado.start(), resultado.end()))


print("\n")

print("Reexecutando o finditer resultados: ")
for resultado in resultados:
    print("\t%s | %s | %s [%s,%s]" % (resultado.group(), resultado.group(1), resultado.group(2), resultado.start(), resultado.end()))

print("\n")

print("Reexecutando o finditer resultados após nova atribuição: ")
resultados = re.finditer(regex, alvo) # Retorna um iterador.
for resultado in resultados:
    print("\t%s | %s | %s [%s,%s]" % (resultado.group(), resultado.group(1), resultado.group(2), resultado.start(), resultado.end()))
